presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java

```diff
@@ -220,6 +230,145 @@ public void testYield()
+                    checkState(input.hasNext(), "spill requested DURING_USAGE, but input is kind of exhausted");
+                    triggerMemoryRevokingAndWait(buildSideSetup.buildOperators.get(i));
+                }
+            }
+
+            actualPages.addAll(toPages(joinOperator, input));
+
+            // expected
+            MaterializedResult expected = MaterializedResult.resultBuilder(taskContext.getSession(), concat(probePages.getTypesWithoutHash(), buildPages.getTypesWithoutHash()))
+                    .row("20", 123_020L, "20", 200L)
+                    .row("21", 123_021L, "21", 201L)
+                    .row("22", 123_022L, "22", 202L)
+                    .row("23", 123_023L, "23", 203L)
+                    .row("30", 123_000L, "30", 300L)
+                    .row("31", 123_001L, "31", 301L)
+                    .row("32", 123_002L, "32", 302L)
+                    .row("33", 123_003L, "33", 303L)
+                    .build();
+
+            List<Type> types = joinOperator.getTypes();
+            if (probePages.getHashChannel().isPresent()) {
+                List<Integer> hashChannels = ImmutableList.of(probePages.getHashChannel().get());
+                actualPages = dropChannel(actualPages, hashChannels);
+                types = without(types, hashChannels);
+            }
+
+            MaterializedResult actual = OperatorAssertion.toMaterializedResult(joinOperator.getOperatorContext().getSession(), types, actualPages);
+            assertEqualsIgnoreOrder(actual.getMaterializedRows(), expected.getMaterializedRows());
+        }
+    }
+
+    private enum WhenSpill
+    {
+        DURING_BUILD, AFTER_BUILD, DURING_USAGE, NEVER
+    }
+
+    @DataProvider
+    public Object[][] testInnerJoinWithSpillDataProvider()
+    {
+        List<Object[]> result = new ArrayList<>();
+        for (boolean probeHashEnabled : ImmutableList.of(false, true)) {
+            for (WhenSpill whenSpill : WhenSpill.values()) {
+                // spill all
+                result.add(new Object[] {probeHashEnabled, nCopies(PARTITION_COUNT, whenSpill)});
+
+                if (whenSpill != WhenSpill.NEVER) {
+                    // spill one
+                    result.add(new Object[] {probeHashEnabled, concat(singletonList(whenSpill), nCopies(PARTITION_COUNT - 1, WhenSpill.NEVER))});
+                }
+            }
+
+            result.add(new Object[] {probeHashEnabled, concat(asList(WhenSpill.DURING_BUILD, WhenSpill.AFTER_BUILD), nCopies(PARTITION_COUNT - 2, WhenSpill.NEVER))});
+            result.add(new Object[] {probeHashEnabled, concat(asList(WhenSpill.DURING_BUILD, WhenSpill.DURING_USAGE), nCopies(PARTITION_COUNT - 2, WhenSpill.NEVER))});
+        }
+
+        return result.stream().toArray(Object[][]::new);
+    }
+
+    private static void revokeMemory(Operator operator)
+    {
+        getFutureValue(operator.startMemoryRevoke());
+        operator.finishMemoryRevoke();
```

Comment:
Can we validate that something happened?  Specifically, can we have an `@VisibleForTesting` state getter that proves that revocation worked?